File: src/latex2word/transforms/crossref.py
"""Cross-reference resolution pass (IR -> IR).

Implements the PRD's headline differentiator plumbing: collect every labelled
target, assign it a sanitized Word bookmark name and a ``SEQ`` counter, then
rewrite each ``Ref`` to point at the matching bookmark with the right kind. The
back-end then emits live ``SEQ``/``REF``/``PAGEREF`` fields -- numbers that
auto-renumber in Word.
"""

from __future__ import annotations

import re

from .. import ir
from ..report import ConversionReport

_COUNTER = {
    "equation": "Equation",
    "figure": "Figure",
    "table": "Table",
    "section": "Section",
    "algorithm": "Algorithm",
}


def sanitize_bookmark(key: str) -> str:
    """Word bookmark names: start with a letter, no spaces, <= 40 chars."""
    cleaned = re.sub(r"[^A-Za-z0-9_]", "_", key)
    if not cleaned or not cleaned[0].isalpha():
        cleaned = "ref_" + cleaned
    return cleaned[:40]


def _label_kind(block: ir.Block) -> str:
    if isinstance(block, ir.MathBlock):
        return "equation"
    if isinstance(block, ir.Figure):
        return "figure"
    if isinstance(block, ir.Table):
        return "table"
    if isinstance(block, ir.Heading):
        return "section"
    if isinstance(block, ir.Theorem):
        return "theorem"
    if isinstance(block, ir.Algorithm):
        return "algorithm"
    return "generic"


def figure_bookmark(fig: ir.Figure) -> str | None:
    """The bookmark for a figure's number (shared by parent and sub-labels)."""
    if fig.label:
        return sanitize_bookmark(fig.label)
    for sub in fig.subfigures:
        if sub.label:
            return sanitize_bookmark("fig_" + sub.label)
    return None


def _collect(blocks: list[ir.Block], labels: dict[str, ir.LabelInfo]) -> None:
    for block in blocks:
        label = getattr(block, "label", None)
        if label:
            kind = _label_kind(block)
            if isinstance(block, ir.Theorem):
                counter_name = block.counter or "Item"
            else:
                counter_name = _COUNTER.get(kind, "Item")
            labels[label] = ir.LabelInfo(
                kind=kind,  # type: ignore[arg-type]
                counter_name=counter_name,
                bookmark=sanitize_bookmark(label),
            )
        # sub-figure labels resolve to the parent figure's number.
        if isinstance(block, ir.Figure) and block.subfigures:
            bookmark = figure_bookmark(block)
            if bookmark is not None:
                if block.label:
                    labels[block.label] = ir.LabelInfo(
                        kind="figure", counter_name="Figure", bookmark=bookmark
                    )
                for sub in block.subfigures:
                    if sub.label:
                        labels[sub.label] = ir.LabelInfo(
                            kind="figure", counter_name="Figure", bookmark=bookmark
                        )
        # recurse into nested block containers
        if isinstance(block, ir.Quote | ir.Theorem):
            _collect(block.blocks, labels)
        elif isinstance(block, ir.ItemList):
            for item in block.items:
                _collect(item.blocks, labels)


def _rewrite_refs_inlines(inlines: list[ir.Inline], labels, report) -> None:
    for node in inlines:
        if isinstance(node, ir.Ref):
            info = labels.get(node.key)
            if info is None:
                report.warn("\\ref", f"unresolved reference to '{node.key}'")
                continue
            node.bookmark = info.bookmark
            if node.ref_kind == "generic":
                node.ref_kind = info.kind  # type: ignore[assignment]
        elif isinstance(node, ir.Emphasis | ir.Link | ir.Footnote | ir.Colored | ir.FontSize):
            _rewrite_refs_inlines(node.inlines, labels, report)


def _rewrite_refs_blocks(blocks: list[ir.Block], labels, report) -> None:
    for block in blocks:
        if isinstance(block, ir.Paragraph | ir.Heading):
            _rewrite_refs_inlines(block.inlines, labels, report)
        elif isinstance(block, ir.Quote | ir.Theorem):
            _rewrite_refs_blocks(block.blocks, labels, report)
        elif isinstance(block, ir.ItemList):
            for item in block.items:
                _rewrite_refs_blocks(item.blocks, labels, report)
        elif isinstance(block, ir.Table) and block.caption:
            _rewrite_refs_inlines(block.caption, labels, report)
        elif isinstance(block, ir.Figure) and block.caption:
            _rewrite_refs_inlines(block.caption, labels, report)


def resolve_crossrefs(doc: ir.Document, report: ConversionReport) -> ir.Document:
    """Populate ``doc.labels`` and wire ``Ref`` nodes to bookmarks (in place)."""
    labels: dict[str, ir.LabelInfo] = {}
    _collect(doc.blocks, labels)
    if doc.meta.abstract:
        _collect(doc.meta.abstract, labels)
    doc.labels = labels
    _rewrite_refs_blocks(doc.blocks, labels, report)
    if doc.meta.abstract:
        _rewrite_refs_blocks(doc.meta.abstract, labels, report)
    return doc

File: rust/crates/tex2word/src/crossref.rs
//! Cross-reference resolution pass (IR → IR) — a Rust port of the Python
//! `transforms/crossref.py`.
//!
//! Two walks over the document: first collect every `\label` target, giving each
//! a sanitized Word bookmark and a `SEQ` counter name; then rewrite every
//! [`Inline::Ref`] to carry the matching bookmark (and inherit the target's kind
//! for a generic `\ref`). `\nameref` becomes an internal hyperlink to the
//! target's title text; `\hyperref[label]` anchors resolve to the sanitized
//! bookmark. Unresolved references are reported, not fatal.

use std::collections::HashMap;

use tex2word_ir::{Block, Document, Inline, LabelInfo, RefKind};

/// A non-fatal diagnostic collected during conversion.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Warning {
    pub context: String,
    pub message: String,
}

/// Resolve cross-references in place: populate `doc.labels` and wire every
/// `Ref`/`Link` to its bookmark. Returns any unresolved-reference warnings.
pub fn resolve(doc: &mut Document) -> Vec<Warning> {
    let mut labels: HashMap<String, LabelInfo> = HashMap::new();
    collect(&doc.blocks, &mut labels);
    doc.labels = labels;
    let mut warnings = Vec::new();
    // Take the labels out to avoid borrowing doc immutably + mutably at once.
    let labels = std::mem::take(&mut doc.labels);
    rewrite_blocks(&mut doc.blocks, &labels, &mut warnings);
    doc.labels = labels;
    warnings
}

/// Word bookmark names must start with a letter, contain no spaces, and be
/// ≤ 40 chars.
pub fn sanitize_bookmark(key: &str) -> String {
    let mut cleaned: String = key
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '_' {
                c
            } else {
                '_'
            }
        })
        .collect();
    if cleaned
        .chars()
        .next()
        .is_none_or(|c| !c.is_ascii_alphabetic())
    {
        cleaned = format!("ref_{cleaned}");
    }
    cleaned.chars().take(40).collect()
}

/// The `SEQ` counter name for a reference kind (mirrors Python `_COUNTER`).
fn counter_name(kind: RefKind) -> &'static str {
    match kind {
        RefKind::Equation => "Equation",
        RefKind::Figure => "Figure",
        RefKind::Table => "Table",
        RefKind::Section => "Section",
        RefKind::Theorem => "Theorem",
        _ => "Item",
    }
}

/// Flatten an inline list to plain text (for a `\nameref` display string).
fn inline_text(inlines: &[Inline]) -> String {
    let mut out = String::new();
    for n in inlines {
        match n {
            Inline::Text(t) => out.push_str(t),
            Inline::Emphasis { inlines, .. } | Inline::Link { inlines, .. } => {
                out.push_str(&inline_text(inlines))
            }
            _ => {}
        }
    }
    out.trim().to_string()
}

fn add_label(
    labels: &mut HashMap<String, LabelInfo>,
    key: &Option<String>,
    kind: RefKind,
    name: Option<String>,
) {
    let Some(key) = key else { return };
    labels.insert(
        key.clone(),
        LabelInfo {
            kind,
            counter_name: counter_name(kind).to_string(),
            bookmark: sanitize_bookmark(key),
            name,
        },
    );
}

fn collect(blocks: &[Block], labels: &mut HashMap<String, LabelInfo>) {
    for block in blocks {
        match block {
            Block::Heading { inlines, label, .. } => {
                add_label(labels, label, RefKind::Section, Some(inline_text(inlines)));
            }
            Block::MathBlock { label, .. } => {
                add_label(labels, label, RefKind::Equation, None);
            }
            Block::Float(f) => {
                let kind = match f.kind {
                    tex2word_ir::FloatKind::Figure => RefKind::Figure,
                    tex2word_ir::FloatKind::Table => RefKind::Table,
                };
                let name = f.caption.as_ref().map(|c| inline_text(c));
                add_label(labels, &f.label, kind, name);
                collect(&f.content, labels); // nested tables/labels
            }
            Block::Quote(bs) => collect(bs, labels),
            _ => {}
        }
    }
}

fn rewrite_blocks(
    blocks: &mut [Block],
    labels: &HashMap<String, LabelInfo>,
    warns: &mut Vec<Warning>,
) {
    for block in blocks {
        match block {
            Block::Heading { inlines, .. } | Block::Paragraph { inlines } => {
                rewrite_inlines(inlines, labels, warns)
            }
            Block::Quote(bs) => rewrite_blocks(bs, labels, warns),
            Block::Float(f) => {
                if let Some(cap) = &mut f.caption {
                    rewrite_inlines(cap, labels, warns);
                }
                rewrite_blocks(&mut f.content, labels, warns);
            }
            Block::Table(t) => {
                for row in &mut t.rows {
                    for cell in &mut row.cells {
                        rewrite_inlines(&mut cell.inlines, labels, warns);
                    }
                }
            }
            _ => {}
        }
    }
}

fn rewrite_inlines(
    inlines: &mut [Inline],
    labels: &HashMap<String, LabelInfo>,
    warns: &mut Vec<Warning>,
) {
    for node in inlines.iter_mut() {
        // \nameref -> an internal hyperlink to the target's title text.
        if let Inline::Ref {
            key,
            kind: RefKind::Name,
            ..
        } = node
        {
            if let Some(info) = labels.get(key) {
                let text = info.name.clone().unwrap_or_else(|| key.clone());
                let anchor = info.bookmark.clone();
                *node = Inline::Link {
                    inlines: vec![Inline::Text(text)],
                    url: String::new(),
                    anchor: Some(anchor),
                };
            } else {
                warns.push(Warning {
                    context: "\\nameref".into(),
                    message: format!("unresolved reference to '{key}'"),
                });
            }
            continue;
        }
        match node {
            Inline::Ref {
                key,
                kind,
                bookmark,
                ..
            } => match labels.get(key) {
                Some(info) => {
                    *bookmark = Some(info.bookmark.clone());
                    if *kind == RefKind::Generic {
                        *kind = info.kind;
                    }
                }
                None => warns.push(Warning {
                    context: "\\ref".into(),
                    message: format!("unresolved reference to '{key}'"),
                }),
            },
            Inline::Link {
                inlines, anchor, ..
            } => {
                if let Some(a) = anchor {
                    match labels.get(a) {
                        Some(info) => *a = info.bookmark.clone(),
                        None => warns.push(Warning {
                            context: "\\hyperref".into(),
                            message: format!("unresolved \\hyperref to '{a}'"),
                        }),
                    }
                }
                rewrite_inlines(inlines, labels, warns);
            }
            Inline::Emphasis { inlines, .. } | Inline::Footnote { inlines } => {
                rewrite_inlines(inlines, labels, warns)
            }
            _ => {}
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tex2word_ir::{Float, FloatKind, RefStyle};

    #[test]
    fn sanitize_rules() {
        assert_eq!(sanitize_bookmark("fig:plot 1"), "fig_plot_1");
        assert_eq!(sanitize_bookmark("2eq"), "ref_2eq"); // must start with a letter
        assert_eq!(sanitize_bookmark("").len(), 4); // "ref_"
    }

    #[test]
    fn collect_and_rewrite() {
        let mut doc = Document {
            blocks: vec![
                Block::Heading {
                    level: 1,
                    inlines: vec![Inline::Text("Intro".into())],
                    label: Some("sec:intro".into()),
                    numbered: true,
                },
                Block::Paragraph {
                    inlines: vec![
                        Inline::Ref {
                            key: "fig:a".into(),
                            kind: RefKind::Generic,
                            style: RefStyle::Plain,
                            bookmark: None,
                        },
                        Inline::Ref {
                            key: "missing".into(),
                            kind: RefKind::Generic,
                            style: RefStyle::Plain,
                            bookmark: None,
                        },
                    ],
                },
                Block::Float(Float {
                    kind: FloatKind::Figure,
                    content: vec![],
                    caption: Some(vec![Inline::Text("A plot".into())]),
                    centered: false,
                    label: Some("fig:a".into()),
                    spanning: false,
                }),
            ],
            ..Default::default()
        };
        let warns = resolve(&mut doc);
        // labels collected
        assert_eq!(doc.labels["sec:intro"].kind, RefKind::Section);
        assert_eq!(doc.labels["fig:a"].bookmark, "fig_a");
        assert_eq!(doc.labels["fig:a"].name.as_deref(), Some("A plot"));
        // generic ref to a figure inherits Figure kind + bookmark
        let Block::Paragraph { inlines } = &doc.blocks[1] else {
            panic!()
        };
        assert!(matches!(&inlines[0],
            Inline::Ref { kind: RefKind::Figure, bookmark: Some(b), .. } if b == "fig_a"));
        // the unresolved ref stays bookmark-less and is reported
        assert!(matches!(&inlines[1], Inline::Ref { bookmark: None, .. }));
        assert_eq!(warns.len(), 1);
        assert!(warns[0].message.contains("missing"));
    }

    #[test]
    fn nameref_becomes_link() {
        let mut doc = Document {
            blocks: vec![
                Block::Heading {
                    level: 1,
                    inlines: vec![Inline::Text("Methods".into())],
                    label: Some("sec:m".into()),
                    numbered: true,
                },
                Block::Paragraph {
                    inlines: vec![Inline::Ref {
                        key: "sec:m".into(),
                        kind: RefKind::Name,
                        style: RefStyle::Plain,
                        bookmark: None,
                    }],
                },
            ],
            ..Default::default()
        };
        resolve(&mut doc);
        let Block::Paragraph { inlines } = &doc.blocks[1] else {
            panic!()
        };
        assert!(matches!(&inlines[0],
            Inline::Link { inlines, anchor: Some(a), .. }
                if a == "sec_m" && inlines == &vec![Inline::Text("Methods".into())]));
    }
}
